### ISSUE TITLE
Sources are downloaded even if not used by the root page anymore

### ISSUE THREAD

**Issue by lapwat** [website-scraper/node-website-scraper#433]
## What I have succeeded to achieve

I am using this plugin to replace images src attributes like so:

```html
<img src="docs/assets/logo.png" />
```

will become

```html
<img src="https://examples.org/docs/assets/logo.png" />
```

## Source code

```js
class WebsiteScraperPlugin {
  apply(registerAction) {
    registerAction('getReference', ({ resource, parentResource, originalReference }) => {
      if (!resource)
        return { reference: parentResource.url + originalReference }; // this line transforms the url of images
      return { reference: resource.url };
    });
  }
}

await scrape({
  urls,
  urlFilter: url => url.startsWith(baseUrl),
  directory,
  sources: [{ selector: 'img', attr: 'src' }],
  plugins: [ new WebsiteScraperPlugin() ],
});
```

## Behavior

It is working as expected, the issue is that **the scraper is downloading the images in the images folder** even if now I have absolute urls.

If I remove `sources: [{ selector: 'img', attr: 'src' }]` from the scrapers option, the plugin will not be called at all and the urls will not be replaced.

# Question

**Is it possible to replace images src attribute without actually downloading the images?**

Thank you very much for your time.

**Comment by s0ph1e** [comment]
Hi @lapwat 👋 

You can try to skip files on downloading with [urlFilter option](https://github.com/website-scraper/node-website-scraper#urlfilter) or customize [saveResource action](https://github.com/website-scraper/node-website-scraper#saveresource) in your plugin.

Hope it helps. 
Please let me know if it doesn't help, I remember I did something similar before but I don't remember the details right now.

**Comment by lapwat** [comment]
Thanks!

I think that saveResource will do the trick. How can I fallback to original saveResource function inside the saveResource action ?

**Comment by lapwat** [comment]
Something like:

```js
registerAction('saveResource', async ({ resource }) => {
  const extension = resource.filename.split('.').pop().toLowerCase();
  const imageExtensions = ['jpg', 'jpeg', 'png'];

  if (imageExtensions.includes(extension)) return
  
  // fallback to original saveResource action
});

```

**Comment by s0ph1e** [comment]
Hi @lapwat 👋 

Not sure there is a good way to access original saveResources, not-so-good way may be just copy implementation from https://github.com/website-scraper/node-website-scraper/blob/525a1df9f771ec0fa0a76bed8d49a166c45a9428/lib/plugins/save-resource-to-fs-plugin.js#L23


I have another idea how to achieve what you want:
* [afterResponse](https://github.com/website-scraper/node-website-scraper#afterresponse) action which returns `null` for resources you don't need to download 
* [getReference](https://github.com/website-scraper/node-website-scraper#getreference) like [in this test](https://github.com/website-scraper/node-website-scraper/blob/5f4bb623154dc5651e2f3c075aac394fd65a04d9/test/functional/update-missing-sources/update-missing-sources.test.js). I believe it may work

Hope at least one of the suggestions helps

**Comment by lapwat** [comment]
The issue here is that the image is downloaded even if not saved. I would like to **prevent downloading AND saving images**.

In `saveResource` and `afterResponse` the image has already been downloaded, it is too late to prevent it.

I tried to use beforeRequest action:

```js
    registerAction('beforeRequest', async ({resource, requestOptions}) => {
      const extension = resource.getFilename().split('.').pop().toLowerCase();
      const imageExtensions = ['jpg', 'jpeg', 'png', 'gif', 'svg'];

      if (imageExtensions.includes(extension))
        return null;                    // prevent download ???

      return {requestOptions};
    });
```

when `beforeRequest` returns null for a resource, getReference is not called for that resource.

It seems like `website-scraper` is not intended to perform this kind of tasks and that I am looking in the wrong direction. If you have more suggestions please share it. Otherwise you can close the issue.

Thanks for your help :)

# Sum up

To sum up what I would like to perform:
- replace relative urls with absolute urls on the root page (ie /images/logo.svg --> https://example/org/images/logo.svg)
- prevent those images to be downloaded and saved

**Comment by s0ph1e** [comment]
@lapwat did you try afterResponse as I suggested in previous comment (not beforeRequest)?

**Comment by lapwat** [comment]
Yes I tries `afterResponse` which where response.body contains the image in binary format. So I guess it has already been downloaded right ?

**Comment by s0ph1e** [comment]
@lapwat no, it means that scraper received http response but didn't save it yet. 
On this step we can skip saving to filesystem if we return null for resource. If object (not null) is returned - scraper will save it, if null - scraper can update the reference in parent resource

**Comment by s0ph1e** [comment]
@lapwat I'm not sure if afterResponse should return null or rejected promise for skipping resource, docs says promise but in example I see null. I suggest to read docs 
https://github.com/website-scraper/node-website-scraper#afterresponse and try both options 🙂

**Comment by lapwat** [comment]
I am not sure we understand each other correctly. If response.body contains data, it means that the data has been downloaded already (even if not saved in a specific file yet).

I would like to optimize the bandwidth of my script so I don't want **the HTTP request to be performed at all**. That is why I think that dealing with afterResponse for my issue is too late.

**Comment by s0ph1e** [comment]
@lapwat aha, got it. Then I suggest to use [urlFilter](https://github.com/website-scraper/node-website-scraper#urlfilter) to filter resources before doing http request and then use [getReference](https://github.com/website-scraper/node-website-scraper#getreference) to update link for missing resources. 

I found an example of exactly same use-case it tests with urlFilter and getReference, please check https://github.com/website-scraper/node-website-scraper/blob/5f4bb623154dc5651e2f3c075aac394fd65a04d9/test/functional/update-missing-sources/update-missing-sources.test.js#L84-L107

**Comment by lapwat** [comment]
Success!

Reference is updated:

```js
class WebsiteScraperPlugin {
  apply(registerAction) {
    registerAction('getReference', ({ parentResource, originalReference }) => {
      const absoluteReference = originalReference.startsWith('http')
        ? originalReference
        : parentResource.url + originalReference;

      console.log('Setting absolute reference for image:', originalReference);
      console.log('->', absoluteReference);

      return { reference: absoluteReference };
    });
  }
}
```

But HTTP request is skipped by url filter:

```js
const imageExtensions = ['jpg', 'jpeg', 'png'];

await scrape({
  urls,
  urlFilter: url => {
    url = new URL(url);
    const extension = url.pathname.split('.').pop().toLowerCase();

    if (imageExtensions.includes(extension)) {
      console.log('Skipping image:', url.pathname); // HTTP request is skipped
      return false;
    }

    return true;
  },
  maxRecursiveDepth: 1,
  directory,
  sources: [{ selector: 'img', attr: 'src' }],
  plugins: [ new WebsiteScraperPlugin() ],
});
```

**`getReference` action happens before `urlFilter`**

Thank you for your patience :)

**Comment by s0ph1e** [comment]
Happy that it finally works 🙂
